DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/compute_units/src/compute_units.cpp

```diff
@@ -0,0 +1,110 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <CL/sycl.hpp>
+#include <iostream>
+
+// dpc_common.hpp can be found in the dev-utilities include folder,
+// e.g., $ONEAPI_ROOT/dev-utilities/latest/include/dpc_common.hpp
+#include "dpc_common.hpp"
+#include "compute_units.hpp"
+#include "pipe_array.hpp"
+
+// Header locations and some DPC++ extensions changed between beta09 and beta10
+// Temporarily modify the code sample to accept either version
+#define BETA09 20200827
+#if __SYCL_COMPILER_VERSION <= BETA09
+  #include <CL/sycl/intel/fpga_extensions.hpp>
+  namespace INTEL = sycl::intel;  // Namespace alias for backward compatibility
+#else
+  #include <CL/sycl/INTEL/fpga_extensions.hpp>
+#endif
+
+
+using namespace sycl;
+
+constexpr float kTestData = 555;
+constexpr size_t kEngines = 5;
+
+using Pipes = PipeArray<class MyPipe, float, 1, kEngines + 1>;
+
+// Forward declaration of the kernel name
+// (This will become unnecessary in a future compiler version.)
+class Source;
+class Sink;
+template <std::size_t ID> class ChainComputeUnit;
+
+// Write the data into the chain
+void SourceKernel(queue &q, float data) {
+
+  q.submit([&](handler &h) {
+    h.single_task<Source>([=] { Pipes::PipeAt<0>::write(data); });
+  });
+}
+
+// Get the data out of the chain and return it to the host
+void SinkKernel(queue &q, std::array<float, 1> &out_data) {
+
+  // Use verbose SYCL 1.2 syntax for the output buffer.
+  // (This will become unnecessary in a future compiler version.)
+  buffer<float, 1> out_buf(out_data.data(), 1);
```

Comment:
Probably
```c++
   buffer out_buf(out_data.data(), 1);
```

Reply:
There was a copy-back bug in an earlier compiler version that prevented the use of the concise syntax. I believe that it is fixed in beta09, but please test functionality in both beta09 and the WIP beta10 compiler if you make this change.

Reply:
```c++
buffer out_buf(out_data.data(), 1);
```
won't work, but this will
```c++
buffer out_buf(out_data);
```

Reply:
Yes, that's the preferred syntax. I checked -- the issue that prevented its use is resolved in beta09.

All code samples that implemented the beta08 workaround should be updated to use the concise syntax.